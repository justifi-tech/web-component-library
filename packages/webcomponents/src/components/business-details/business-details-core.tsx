import { Component, Event, EventEmitter, Prop, State, h } from '@stencil/core';
import { Business } from '../../api/Business';
import { ErrorState } from '../../ui-components/details/utils';
import { ComponentError } from '../../api/ComponentError';
import { Skeleton, StyledHost } from '../../ui-components';

enum RENDER_STATES {
  LOADING = 'loading',
  READY = 'ready',
  ERROR = 'error',
}

@Component({
  tag: 'business-details-core'
})
export class BusinessDetailsCore {
  @Prop() getBusiness: Function;

  @State() business: Business;
  @State() renderState: RENDER_STATES = RENDER_STATES.LOADING;
  @State() errorMessage: string = 'An error ocurred.';

  @Event({ eventName: 'error-event' }) errorEvent: EventEmitter<ComponentError>;

  async componentWillLoad() {
    if (this.getBusiness) {
      this.fetchData();
    }
  }

  fetchData() {
    this.renderState = RENDER_STATES.LOADING;

    this.getBusiness({
      onSuccess: ({ business }) => {
        this.business = business;
        this.renderState = RENDER_STATES.READY;
      },
      onError: ({ error, code, severity }) => {
        this.errorMessage = error;
        this.renderState = RENDER_STATES.ERROR;

        this.errorEvent.emit({
          errorCode: code,
          message: error,
          severity,
        });
      }
    });
  }

  render() {
    if (this.renderState === RENDER_STATES.LOADING) {
      return (
        <StyledHost>
          <Skeleton height="100px" />
          <Skeleton height="100px" />
          <Skeleton height="100px" />
        </StyledHost>
      );
    }

    if (this.renderState === RENDER_STATES.ERROR) {
      return <StyledHost>{ErrorState(this.errorMessage)}</StyledHost>;
    }

    return (
      <StyledHost>
        <generic-info-details business={this.business} />
        <legal-address-details legalAddress={this.business.legal_address} />
        <representative-details representative={this.business.representative} />
        <owner-details owners={this.business.owners} />
        <additional-questions-details
          additionalQuestions={this.business.additional_questions}
        />
      </StyledHost>
    );
  }
}
