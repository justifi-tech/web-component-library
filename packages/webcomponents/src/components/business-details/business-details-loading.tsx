import { Fragment, h } from "@stencil/core";
import { Skeleton } from "../../ui-components";

export const BusinessDetailsLoading = () => {
  const Block = () => (
    <div class="mt-5">
      <Skeleton width="40%" height="1.5rem" />
      <hr />

      <div class="row g-3">
        <div class="col-12 col-md-6">
          <Skeleton />
          <Skeleton />
          <Skeleton />
        </div>
        <div class="col-12 col-md-6">
          <Skeleton />
          <Skeleton />
          <Skeleton />
        </div>
      </div>
    </div>
  );

  return (
    <Fragment>
      <Block />
      <Block />
      <Block />
      <Block />
    </Fragment>
  );
}
