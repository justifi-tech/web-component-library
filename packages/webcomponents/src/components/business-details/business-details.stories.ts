import { config } from '../../../config';

export default {
  title: 'Components/BusinessDetails',
  component: 'justifi-business-details',
  parameters: {},
};

class BusinessDetailsArgs {
  'businessId': string;
  'authToken': string;

  constructor(args) {
    this['businessId'] = args['businessId'] || config.exampleBusinessId;
    this['authToken'] = args['authToken'] || config.proxyAuthToken;
  }
};

const Template = (args: BusinessDetailsArgs) => {
  return `
    <justifi-business-details
      business-id="${args['businessId']}"
      auth-token="${args['authToken']}"
    />
  `;
};

export const Basic = Template.bind({});
Basic.args = new BusinessDetailsArgs({});
