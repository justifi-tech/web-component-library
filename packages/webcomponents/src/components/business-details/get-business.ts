import { Business } from '../../api/Business';
import { getErrorMessage } from '../../api/services/utils';

export const makeGetBusiness =
  ({ id, authToken, service }) =>
  async ({ onSuccess, onError }) => {
    try {
      const response = await service.fetchBusiness(id, authToken);

      if (!response.error) {
        onSuccess({ business: new Business(response.data) });
      } else {
        const responseError = getErrorMessage(response.error);
        return onError({ error: responseError });
      }
    } catch (error) {
      return onError({ error: error.message || error });
    }
  };
