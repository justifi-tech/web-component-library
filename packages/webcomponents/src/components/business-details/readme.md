# justifi-business-details



<!-- Auto Generated Below -->


## Properties

| Property     | Attribute     | Description | Type     | Default     |
| ------------ | ------------- | ----------- | -------- | ----------- |
| `authToken`  | `auth-token`  |             | `string` | `undefined` |
| `businessId` | `business-id` |             | `string` | `undefined` |


## Events

| Event         | Description | Type                               |
| ------------- | ----------- | ---------------------------------- |
| `error-event` |             | `CustomEvent<ComponentErrorEvent>` |


## Dependencies

### Depends on

- [business-details-core](.)

### Graph
```mermaid
graph TD;
  justifi-business-details --> business-details-core
  business-details-core --> justifi-details
  business-details-core --> core-info-details
  business-details-core --> legal-address-details
  business-details-core --> representative-details
  business-details-core --> owner-details
  business-details-core --> additional-questions-details
  style justifi-business-details fill:#f9f,stroke:#333,stroke-width:4px
```

----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
