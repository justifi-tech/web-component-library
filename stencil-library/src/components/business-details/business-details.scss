@import 'reboot';
@import 'bootstrap-utilities';
@import 'type';
@import 'spinners';

:host {
  @include host-base-styles;
}
