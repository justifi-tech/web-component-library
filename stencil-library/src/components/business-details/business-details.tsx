import { Component, Host, Prop, State, h } from '@stencil/core';
import { IBusiness } from '../../api/Business';
import { Api } from '../../api';
import { ErrorState, LoadingState } from '../details/utils'; // Make sure to adjust the path if necessary

enum RENDER_STATES {
  LOADING = 'loading',
  READY = 'ready',
  ERROR = 'error',
}

/**
 *
 * @exportedPart detail-section
 * @exportedPart detail-section-title
 * @exportedPart detail-section-item-title
 * @exportedPart detail-section-item-data
 * @exportedPart detail-empty-state
 */
@Component({
  tag: 'justifi-business-details',
  styleUrl: 'business-details.scss',
  shadow: true,
})
export class BusinessDetails {
  @Prop() businessId: string;
  @Prop() authToken: string;
  @State() business: IBusiness;
  @State() renderState: RENDER_STATES = RENDER_STATES.LOADING;
  @State() errorMessage: string = 'An error ocurred.';

  private api: any;

  constructor() {
    this.fetchBusiness = this.fetchBusiness.bind(this);
    this.setError = this.setError.bind(this);
  }

  async componentWillLoad() {
    if (!this.authToken) {
      this.setError(
        'Missing auth-token. The form will not be functional without it.',
      );
      return;
    }

    if (!this.businessId) {
      this.setError(
        'Missing business-id. The form will not be functional without it.',
      );
      return;
    }

    this.api = Api(this.authToken, process.env.ENTITIES_ENDPOINT);
    await this.fetchBusiness(this.businessId);
  }

  async fetchBusiness(businessId) {
    this.renderState = RENDER_STATES.LOADING;
    try {
      const response = await this.api.get(`entities/business/${businessId}`);
      if (response.error) {
        this.setError(response.error.message);
        return;
      }
      this.business = response.data;
      this.renderState = RENDER_STATES.READY;
    } catch (error) {
      this.setError(error);
    }
  }

  setError(error: string) {
    this.errorMessage = error || this.errorMessage;
    this.renderState = RENDER_STATES.ERROR;
  }

  render() {
    if (this.renderState === RENDER_STATES.LOADING) {
      return LoadingState;
    }

    if (this.renderState === RENDER_STATES.ERROR) {
      return ErrorState(this.errorMessage);
    }

    return (
      <Host>
        <business-details-section business={this.business} />
        <representative-details-section business={this.business} />
        <owner-details-section business={this.business} />
      </Host>
    );
  }
}
