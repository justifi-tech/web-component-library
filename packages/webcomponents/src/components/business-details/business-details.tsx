import { Component, Event, EventEmitter, Host, Prop, State, h } from '@stencil/core';
import { ErrorState } from '../../ui-components/details/utils';
import { BusinessService } from '../../api/services/business.service';
import { makeGetBusiness } from './get-business';
import { ComponentError, ComponentErrorCodes, ComponentErrorSeverity } from '../../api/ComponentError';
import JustifiAnalytics from '../../api/Analytics';

/**
 *
 * @exportedPart detail-section
 * @exportedPart detail-section-title
 * @exportedPart detail-section-item-title
 * @exportedPart detail-section-item-data
 * @exportedPart detail-empty-state
 */
@Component({
  tag: 'justifi-business-details',
  shadow: true,
})
export class BusinessDetails {
  @Prop() businessId: string;
  @Prop() authToken: string;

  @State() errorMessage: string;
  @State() getBusiness: Function;

  analytics: JustifiAnalytics;

  @Event({
    eventName: 'error-event'
  }) errorEvent: EventEmitter<ComponentError>;


  componentWillLoad() {
    this.analytics = new JustifiAnalytics(this);
    this.initializeGetBusiness();
  }

  disconnectedCallback() {
    this.analytics.cleanup();
  }

  private initializeGetBusiness() {
    if (!this.businessId || !this.authToken) {
      this.errorMessage = 'Invalid business id or auth token';
      this.errorEvent.emit({
        errorCode: ComponentErrorCodes.MISSING_PROPS,
        message: this.errorMessage,
        severity: ComponentErrorSeverity.ERROR,
      });
      return;
    }

    this.getBusiness = makeGetBusiness({
      id: this.businessId,
      authToken: this.authToken,
      service: new BusinessService(),
    });
  }

  render() {
    if (this.errorMessage) {
      return <Host>{ErrorState(this.errorMessage)}</Host>;
    }
    return (
      <Host>
        <business-details-core getBusiness={this.getBusiness} />
      </Host>
    );
  }
}
