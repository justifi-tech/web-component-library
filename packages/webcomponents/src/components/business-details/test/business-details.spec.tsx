import { h } from '@stencil/core';
import { newSpecPage } from '@stencil/core/testing';
import { BusinessDetails } from '../business-details';
import { BusinessDetailsCore } from '../business-details-core';

describe('justifi-business-details', () => {
  it('initializes getBusiness on load with valid props', async () => {
    const page = await newSpecPage({
      components: [BusinessDetails],
      template: () => <justifi-business-details business-id="123" auth-token="token" />,
    });

    page.waitForChanges();

    expect(page.rootInstance.getBusiness).toBeDefined();
    expect(page.rootInstance.errorMessage).toBeUndefined();
  });

  it('sets an error message with invalid props', async () => {
    const page = await newSpecPage({
      components: [BusinessDetails],
      template: () => <justifi-business-details />, // No props provided
    });

    page.waitForChanges();

    expect(page.rootInstance.errorMessage).toBe('Invalid business id or auth token');
  });

  it('renders the error state when there is an error message', async () => {
    const page = await newSpecPage({
      components: [BusinessDetails, BusinessDetailsCore],
      template: () => <justifi-business-details />,
    });

    await page.waitForChanges();

    expect(page.rootInstance.errorMessage).toBe('Invalid business id or auth token');
    expect(page.root).toMatchSnapshot();
  });

  it('renders business-details-core when there is no error message', async () => {
    const page = await newSpecPage({
      components: [BusinessDetails, BusinessDetailsCore],
      template: () => <justifi-business-details business-id="123" auth-token="token" />,
    });

    page.waitForChanges();

    const core = page.root.shadowRoot.querySelector('business-details-core');
    expect(core).not.toBeNull();
  });

  it('emits an error event when there is no auth token', async () => {
    const errorSpy = jest.fn();

    const page = await newSpecPage({
      components: [BusinessDetails],
      template: () => <justifi-business-details business-id="123" onErrorEvent={errorSpy} />,
    });

    page.waitForChanges();

    expect(errorSpy).toHaveBeenCalledWith(
      expect.objectContaining({
        detail: {
          errorCode: 'missing-props',
          message: 'Invalid business id or auth token',
          severity: 'error',
        }
      })
    );
  });
});
