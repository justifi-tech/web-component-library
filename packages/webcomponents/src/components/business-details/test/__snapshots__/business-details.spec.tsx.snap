// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`justifi-business-details renders the error state when there is an error message 1`] = `
<justifi-business-details>
  <template shadowrootmode="open">
    <main class="d-flex justify-content-center p-4 text-center" part="detail-empty-state" style="font-size: 1.2rem;">
      <div>
        Invalid business id or auth token
      </div>
    </main>
  </template>
</justifi-business-details>
`;
